### PR TITLE
[Improving Documentation] Issue #110

### LINKED ISSUE #110: Improve Documentation
Currently, our readthedocs documentation is very lack luster. The formatting for some of the function docs looks pretty bad, and the layout for the sidebar needs improvement. We also need example code and some basic tutorials to help people get started (these can possibly be taken from the existing README).

### PR DESCRIPTION
**Pull request by iiskakov**
Fixes #110

* Sidebar looks a little bit better
* Function formatting is also a bit more consistent
* Logo added
* Tutorials added
* "How to contribute" info added

_This change requires a documentation update_

**readthedocs.com build log**:
```
git remote set-url origin https://github.com/iiskakov/TensorNetwork.git
git fetch origin --force --tags --prune --prune-tags --depth 50
git checkout --force origin/master
git clean -d -f -f
python3.7 -mvirtualenv --no-site-packages --no-download /home/docs/checkouts/readthedocs.org/user_builds/42-robotics-tensornetwork/envs/latest
/home/docs/checkouts/readthedocs.org/user_builds/42-robotics-tensornetwork/envs/latest/bin/python -m pip install --upgrade --cache-dir /home/docs/checkouts/readthedocs.org/user_builds/42-robotics-tensornetwork/.cache/pip pip
/home/docs/checkouts/readthedocs.org/user_builds/42-robotics-tensornetwork/envs/latest/bin/python -m pip install --upgrade --cache-dir /home/docs/checkouts/readthedocs.org/user_builds/42-robotics-tensornetwork/.cache/pip Pygments==2.3.1 setuptools==41.0.1 docutils==0.14 mock==1.0.1 pillow==5.4.1 alabaster>=0.7,<0.8,!=0.7.5 commonmark==0.8.1 recommonmark==0.5.0 sphinx<2 sphinx-rtd-theme<0.5 readthedocs-sphinx-ext<1.1
/home/docs/checkouts/readthedocs.org/user_builds/42-robotics-tensornetwork/envs/latest/bin/python -m pip install --exists-action=w --cache-dir /home/docs/checkouts/readthedocs.org/user_builds/42-robotics-tensornetwork/.cache/pip -r requirements.txt
cat docs/conf.py
python /home/docs/checkouts/readthedocs.org/user_builds/42-robotics-tensornetwork/envs/latest/bin/sphinx-build -T -E -b readthedocs -d _build/doctrees-readthedocs -D language=en . _build/html
python /home/docs/checkouts/readthedocs.org/user_builds/42-robotics-tensornetwork/envs/latest/bin/sphinx-build -T -b readthedocssinglehtmllocalmedia -d _build/doctrees-readthedocssinglehtmllocalmedia -D language=en . _build/localmedia
python /home/docs/checkouts/readthedocs.org/user_builds/42-robotics-tensornetwork/envs/latest/bin/sphinx-build -b latex -D language=en -d _build/doctrees . _build/latex
cat latexmkrc
latexmk -r latexmkrc -pdf -f -dvi- -ps- -jobname=42-robotics-tensornetwork -interaction=nonstopmode
mv -f /home/docs/checkouts/readthedocs.org/user_builds/42-robotics-tensornetwork/checkouts/latest/docs/_build/latex/42-robotics-tensornetwork.pdf /home/docs/checkouts/readthedocs.org/user_builds/42-robotics-tensornetwork/artifacts/latest/sphinx_pdf/42-robotics-tensornetwork.pdf
python /home/docs/checkouts/readthedocs.org/user_builds/42-robotics-tensornetwork/envs/latest/bin/sphinx-build -T -b epub -d _build/doctrees-epub -D language=en . _build/epub
mv -f /home/docs/checkouts/readthedocs.org/user_builds/42-robotics-tensornetwork
```

### CODE REVIEW

**Comment by iiskakov** [comment]
@googlebot I signed it!

**Comment by iiskakov** [comment]
Update: Fixed long line.

**Comment by Thenerdstation** [comment]
Looks pretty good. Heads up though, since this is such a large PR it might take a few days to fully go through it.

File: docs/mps.rst [hunk 1]
```diff
@@ -1,17 +1,12 @@
...
+     :toctree: stubs
 
-.. autoclass:: tensornetwork.matrixproductstates.finite_mps.FiniteMPS
-    :members:
-       
-.. autoclass:: tensornetwork.matrixproductstates.infinite_mps.InfiniteMPS
-    :members:
+     tensornetwork.matrixproductstates.base_mps.BaseMPS
+     tensornetwork.matrixproductstates.finite_mps.FiniteMPS
+     tensornetwork.matrixproductstates.infinite_mps.InfiniteMPS
```
Comment (Thenerdstation) [inline docs/mps.rst]: Can you remove the `matrixproductstates.{base_mps, finite_mps,...}`? This _should_ just work with `tensornetwork.BaseMPS` etc.

File: docs/network_components.rst [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+.. autofunction:: tensornetwork.outer_product
+.. autofunction:: tensornetwork.outer_product_final_nodes
+.. autofunction:: tensornetwork.contract_parallel
+.. autofunction:: tensornetwork.flatten_edges
+.. autofunction:: tensornetwork.get_all_nondangling
+.. autofunction:: tensornetwork.get_all_dangling
+.. autofunction:: tensornetwork.flatten_all_edges
+.. autofunction:: tensornetwork.flatten_edges_between
+.. autofunction:: tensornetwork.get_parallel_edges
+.. autofunction:: tensornetwork.get_shared_edges
```
Comment (Thenerdstation) [inline docs/network_components.rst]: I wouldn't separate network components and network operations. There is not difference from the users perspective. Can you combine the two files?

File: docs/network_operations.rst [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+.. autofunction:: tensornetwork.reachable
+.. autofunction:: tensornetwork.check_connected
+.. autofunction:: tensornetwork.check_correct
+.. autofunction:: tensornetwork.get_all_nodes
+.. autofunction:: tensornetwork.get_all_edges
+.. autofunction:: tensornetwork.get_subgraph_dangling
+.. autofunction:: tensornetwork.reduced_density
+.. autofunction:: tensornetwork.switch_backend
+.. autofunction:: tensornetwork.remove_node
+.. autofunction:: tensornetwork.contract_trace_edges
```
Comment (Thenerdstation) [inline docs/network_operations.rst]: see above

File: docs/nodes.rst [hunk 4]
```diff
@@ -1,15 +1,8 @@
...
+     :toctree: stubs
 
-.. autoclass:: tensornetwork.Node
-    :members:
-
-.. autoclass:: tensornetwork.CopyNode
-    :members:
-
-.. autoclass:: tensornetwork.BaseNode
-    :members:
```
Comment (Thenerdstation) [inline docs/nodes.rst]: Don't forget to include `CopyNode` and `BaseNode`.

File: tensornetwork/ncon_interface.py [hunk 5]
```diff
@@ -35,45 +36,48 @@ def ncon(tensors: Sequence[Union[network_components.BaseNode, Tensor]],
 
     If a contraction order `con_order` and an output order `out_order`
     are both provided, the edge labels can be anything.
-    Otherwise (`con_order == None or out_order == None`), the edge labels 
+    Otherwise (``con_order == None or out_order == None``), the edge labels
```
Comment (Thenerdstation) [inline tensornetwork/ncon_interface.py]: Can you be sure to only use one grave accent (`) ? Using 2 changes the color of the display to red and it looks kinda bad.

See this for example. The red color are the ones with 2 and the black ones use just 1. 
image

File: tensornetwork/network_operations.py [hunk 6]
```diff
@@ -29,13 +29,16 @@
...
   Args:
     node: A `BaseNode`. 
+
   Returns:
-    The L2 norm 
+    The L2 norm.
+
   Raises:
-    AttributeError: If `node` has no `backend` attribute.
+    AttributeError: If `Node` has no `backend` attribute.
```
Comment (Thenerdstation) [inline tensornetwork/network_operations.py]: since `node` is an argument here, don't capitalize it.

Comment (Thenerdstation) [reply tensornetwork/network_operations.py]: This appears several places. Be sure to update all of them!

File: docs/index.rst [hunk 7]
```diff
@@ -1,22 +1,25 @@
...
 
 .. toctree::
-   :maxdepth: 4
-   :caption: Contents:
+   :maxdepth: 1
+   :caption: Tutorials
 
    tutorial
+   basic_api_tutorial
+   Tensor Networks inside Neural Networks using Keras <https://colab.research.google.com/drive/1JUh84N5sbfQYk6HWowWCGl0IZ1idQi6z>
```
Comment (Thenerdstation) [inline docs/index.rst]: Can you actually include the ipython notebook like you did with the basic API tutorial?

File: docs/mps.rst (hunk 1 above)
Comment (iiskakov) [reply docs/mps.rst]: Done ^^

File: docs/network_components.rst (hunk 2 above)
Comment (iiskakov) [reply docs/network_components.rst]: Done as well

File: docs/nodes.rst (hunk 4 above)
Comment (iiskakov) [reply docs/nodes.rst]: Done, sorry I missed it first time for some reason :<

File: tensornetwork/ncon_interface.py (hunk 5 above)
Comment (iiskakov) [reply tensornetwork/ncon_interface.py]: Done.  ^^

A little follow-up of my motivation before:
I was basically trying to mimic JAX and it's awesome docs, and it seemed to me that their approach is this:
Single accent for referencing or inter-linking something, or just to bring attention
Double accent for small code examples or variable names/data types

File: tensornetwork/network_operations.py (hunk 6 above)
Comment (iiskakov) [reply tensornetwork/network_operations.py]: Done

File: docs/index.rst (hunk 7 above)
Comment (iiskakov) [reply docs/index.rst]: Done as well

File: tensornetwork/ncon_interface.py (hunk 5 above)
Comment (Thenerdstation) [reply tensornetwork/ncon_interface.py]: JAX does have very nice documentation. We should try to mimic them whenever possible.

**Comment by Thenerdstation** [comment]
Awesome! Thanks for the fixes.

Again, this is big so I'll do like 1-2 more passes before I approve.

**Comment by Thenerdstation** [comment]
Wow I dropped the ball on reviewing this. Doing the last look now!

**Pull request merged**